Recompiling causes panic

Making changes to code in modules that compile to WASM and running `cargo build --release` results in a panic.  

This is the project: https://github.com/paritytech/substrate

Immediately running `cargo clean && cargo build --release` will allow successful compilation. Also upgrading to a new version of nightly will allow recompilation without `cargo clean`, but only the first recompile.

## Meta

`rustc --version --verbose`: 

rustc 1.38.0-nightly (6a91782b7 2019-08-06) running on x86_64-apple-darwin

Backtrace:

```
Fresh transaction-factory v0.0.1 (/Users/matt/projects/substrate/test-utils/transaction-factory)
error: failed to run custom build command for `node-runtime v2.0.0 (/Users/matt/projects/substrate/node/runtime)`

Caused by:
  process didn't exit successfully: `/Users/matt/projects/substrate/target/release/build/node-runtime-777af245baf20274/build-script-build` (exit code: 1)
--- stdout
Executing build command: "rustup" "run" "nightly" "cargo" "build" "--target=wasm32-unknown-unknown" "--manifest-path=/Users/matt/projects/substrate/target/release/wbuild/node-runtime/Cargo.toml" "--release"

--- stderr
   Compiling wasm-build-runner-impl v1.0.0 (/Users/matt/projects/substrate/target/release/build/node-runtime-3547c858aedc1441/out/wasm_build_runner)
    Finished release [optimized] target(s) in 0.25s
     Running `/Users/matt/projects/substrate/target/release/build/node-runtime-3547c858aedc1441/out/wasm_build_runner/target/release/wasm-build-runner-impl`
   Compiling node-runtime-wasm v1.0.0 (/Users/matt/projects/substrate/target/release/wbuild/node-runtime)
thread '<unnamed>' panicked at 'must have at least one serialized module', src/libcore/option.rs:1166:5
stack backtrace:
   0:        0x109957f12 - std::panicking::default_hook::{{closure}}::hbefba8356e31122b
   1:        0x109957bdd - std::panicking::default_hook::hb85a284cef21dddc
   2:        0x1085b55a3 - rustc::util::common::panic_hook::h8e13af8cece077e3
   3:        0x1099587d1 - std::panicking::rust_panic_with_hook::h8868b29516eb168e
   4:        0x10995822d - std::panicking::continue_panic_fmt::h8f85da70e3d42ed7
   5:        0x109958119 - rust_begin_unwind
   6:        0x109983e82 - core::panicking::panic_fmt::ha880b2fe6921e68e
   7:        0x109983f9e - core::option::expect_failed::hfea1a2f203365264
   8:        0x10bb851f8 - rustc_codegen_llvm::back::lto::run_fat::h0e6fc34b8aa41a0c
   9:        0x10bc04aa8 - rustc_codegen_ssa::back::write::start_executing_work::{{closure}}::h1c543b9ea0992759
  10:        0x10bc0b7d0 - std::sys_common::backtrace::__rust_begin_short_backtrace::h6b423e28f14e4e74
  11:        0x10bb5279c - std::panicking::try::do_call::h6fca1ea505f089a9
  12:        0x109967b4f - __rust_maybe_catch_panic
  13:        0x10bb52ba6 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h6eb73fcc38df565a
  14:        0x10993bb1e - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h2a90a69cfe4e95fa
  15:        0x10996695e - std::sys::unix::thread::Thread::new::thread_start::h99953f013e830b09
  16:     0x7fff7c8392eb - _pthread_body
  17:     0x7fff7c83c249 - _pthread_start
query stack during panic:
end of query stack
thread 'rustc' panicked at 'src/librustc_codegen_ssa/back/write.rs:1826: panic during codegen/LLVM phase', src/librustc/util/bug.rs:37:26
stack backtrace:
   0:        0x109957f12 - std::panicking::default_hook::{{closure}}::hbefba8356e31122b
   1:        0x109957bdd - std::panicking::default_hook::hb85a284cef21dddc
   2:        0x1085b55a3 - rustc::util::common::panic_hook::h8e13af8cece077e3
   3:        0x1099587d1 - std::panicking::rust_panic_with_hook::h8868b29516eb168e
   4:        0x1085f9b25 - std::panicking::begin_panic::he8bde94d32f67181
   5:        0x10828fbb4 - rustc::util::bug::opt_span_bug_fmt::{{closure}}::h79b523f376a07816
   6:        0x10828f9ea - rustc::ty::context::tls::with_opt::{{closure}}::heef7fe9845124b11
   7:        0x10828f777 - rustc::ty::context::tls::with_context_opt::h61634ca6ab55ac5a
   8:        0x10828f7c2 - rustc::ty::context::tls::with_opt::h5ca54a02c4c0bb9b
   9:        0x10828faf8 - rustc::util::bug::opt_span_bug_fmt::h8f8a983ed3203bff
  10:        0x10828fa4b - rustc::util::bug::bug_fmt::h7fff681b74edf248
  11:        0x10bc09e05 - rustc_codegen_ssa::back::write::OngoingCodegen<B>::join::hd601e779fa22708f
  12:        0x10bc43d98 - <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::join_codegen_and_link::hb5084914450219e0
  13:        0x1070296d2 - rustc_interface::queries::Query<T>::compute::h79dc64dd1d20cdc1
  14:        0x1070ede3c - rustc_interface::queries::<impl rustc_interface::interface::Compiler>::link::hc53a9984fbe3c770
  15:        0x106f5d224 - rustc_interface::interface::run_compiler_in_existing_thread_pool::hbcb4433730475c4a
  16:        0x106f90744 - std::thread::local::LocalKey<T>::with::hfc60cc57f753f9b0
  17:        0x106fa6228 - syntax::with_globals::h974a79eedc8d66fd
  18:        0x106f381a7 - std::sys_common::backtrace::__rust_begin_short_backtrace::ha8d874bf2a6ac95b
  19:        0x109967b4f - __rust_maybe_catch_panic
  20:        0x106f5eeb7 - core::ops::function::FnOnce::call_once{{vtable.shim}}::hf166e26f2620097b
  21:        0x10993bb1e - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h2a90a69cfe4e95fa
  22:        0x10996695e - std::sys::unix::thread::Thread::new::thread_start::h99953f013e830b09
  23:     0x7fff7c8392eb - _pthread_body
  24:     0x7fff7c83c249 - _pthread_start
query stack during panic:
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.38.0-nightly (6a91782b7 2019-08-06) running on x86_64-apple-darwin

note: compiler flags: -C opt-level=3 -C panic=abort -C lto -C incremental -C link-arg=--export-table -C link-arg=--export=__heap_base --crate-type cdylib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `node-runtime-wasm`.
```

Hm I was getting build errors locally and wasn't able to reproduce. I'm not really sure what compiler is necessary to fix errors like:

```
error[E0432]: unresolved import `alloc_crate::collections::CollectionAllocErr`
  --> /home/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/hashmap_core-0.1.11/src/lib.rs:32:13
   |
32 |     pub use alloc_crate::collections::CollectionAllocErr;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `CollectionAllocErr` in `collections`

   Compiling srml-support-procedural-tools v2.0.0 (/home/alex/code/substrate/srml/support/procedural/tools)
```

> Hm I was getting build errors locally and wasn't able to reproduce. I'm not really sure what compiler is necessary to fix errors like:
> 
> ```
> error[E0432]: unresolved import `alloc_crate::collections::CollectionAllocErr`
>   --> /home/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/hashmap_core-0.1.11/src/lib.rs:32:13
>    |
> 32 |     pub use alloc_crate::collections::CollectionAllocErr;
>    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `CollectionAllocErr` in `collections`
> 
>    Compiling srml-support-procedural-tools v2.0.0 (/home/alex/code/substrate/srml/support/procedural/tools)
> ```

I think unfortunately you tried at a time where master was broken! If you can try again, maybe it's possible to reproduce the original problem

--- a/src/librustc_codegen_llvm/back/lto.rs
+++ b/src/librustc_codegen_llvm/back/lto.rs
@@ -183,13 +183,39 @@ pub(crate) fn prepare_thin(
 
 fn fat_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
            diag_handler: &Handler,
-           mut modules: Vec<FatLTOInput<LlvmCodegenBackend>>,
+           modules: Vec<FatLTOInput<LlvmCodegenBackend>>,
            cached_modules: Vec<(SerializedModule<ModuleBuffer>, WorkProduct)>,
            mut serialized_modules: Vec<(SerializedModule<ModuleBuffer>, CString)>,
            symbol_white_list: &[*const libc::c_char])
     -> Result<LtoModuleCodegen<LlvmCodegenBackend>, FatalError>
 {
     info!("going for a fat lto");
+
+    // Sort out all our lists of incoming modules into two lists.
+    //
+    // * `serialized_modules` (also and argument to this function) contains all
+    //   modules that are serialized in-memory.
+    // * `in_memory` contains modules which are already parsed and in-memory,
+    //   such as from multi-CGU builds.
+    //
+    // All of `cached_modules` (cached from previous incremental builds) can
+    // immediately go onto the `serialized_modules` modules list and then we can
+    // split the `modules` array into these two lists.
+    let mut in_memory = Vec::new();
+    serialized_modules.extend(cached_modules.into_iter().map(|(buffer, wp)| {
+        info!("pushing cached module {:?}", wp.cgu_name);
+        (buffer, CString::new(wp.cgu_name).unwrap())
+    }));
+    for module in modules {
+        match module {
+            FatLTOInput::InMemory(m) => in_memory.push(m),
+            FatLTOInput::Serialized { name, buffer } => {
+                info!("pushing serialized module {:?}", name);
+                let buffer = SerializedModule::Local(buffer);
+                serialized_modules.push((buffer, CString::new(name).unwrap()));
+            }
+        }
+    }
 
     // Find the "costliest" module and merge everything into that codegen unit.
     // All the other modules will be serialized and reparsed into the new
@@ -200,14 +226,8 @@ fn fat_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
     // file copy operations in the backend work correctly. The only other kind
     // of module here should be an allocator one, and if your crate is smaller
     // than the allocator module then the size doesn't really matter anyway.
-    let costliest_module = modules.iter()
+    let costliest_module = in_memory.iter()
         .enumerate()
-        .filter_map(|(i, module)| {
-            match module {
-                FatLTOInput::InMemory(m) => Some((i, m)),
-                FatLTOInput::Serialized { .. } => None,
-            }
-        })
         .filter(|&(_, module)| module.kind == ModuleKind::Regular)
         .map(|(i, module)| {
             let cost = unsafe {
@@ -223,26 +243,14 @@ fn fat_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
     // re-executing the LTO passes. If that's the case deserialize the first
     // module and create a linker with it.
     let module: ModuleCodegen<ModuleLlvm> = match costliest_module {
-        Some((_cost, i)) => {
-            match modules.remove(i) {
-                FatLTOInput::InMemory(m) => m,
-                FatLTOInput::Serialized { .. } => unreachable!(),
-            }
-        }
+        Some((_cost, i)) => in_memory.remove(i),
         None => {
-            let pos = modules.iter().position(|m| {
-                match m {
-                    FatLTOInput::InMemory(_) => false,
-                    FatLTOInput::Serialized { .. } => true,
-                }
-            }).expect("must have at least one serialized module");
-            let (name, buffer) = match modules.remove(pos) {
-                FatLTOInput::Serialized { name, buffer } => (name, buffer),
-                FatLTOInput::InMemory(_) => unreachable!(),
-            };
+            assert!(serialized_modules.len() > 0, "must have at least one serialized module");
+            let (buffer, name) = serialized_modules.remove(0);
+            info!("no in-memory regular modules to choose from, parsing {:?}", name);
             ModuleCodegen {
-                module_llvm: ModuleLlvm::parse(cgcx, &name, &buffer, diag_handler)?,
-                name,
+                module_llvm: ModuleLlvm::parse(cgcx, &name, buffer.data(), diag_handler)?,
+                name: name.into_string().unwrap(),
                 kind: ModuleKind::Regular,
             }
         }
@@ -265,25 +273,13 @@ fn fat_lto(cgcx: &CodegenContext<LlvmCodegenBackend>,
         // and we want to move everything to the same LLVM context. Currently the
         // way we know of to do that is to serialize them to a string and them parse
         // them later. Not great but hey, that's why it's "fat" LTO, right?
-        let mut new_modules = modules.into_iter().map(|module| {
-            match module {
-                FatLTOInput::InMemory(module) => {
-                    let buffer = ModuleBuffer::new(module.module_llvm.llmod());
-                    let llmod_id = CString::new(&module.name[..]).unwrap();
-                    (SerializedModule::Local(buffer), llmod_id)
-                }
-                FatLTOInput::Serialized { name, buffer } => {
-                    let llmod_id = CString::new(name).unwrap();
-                    (SerializedModule::Local(buffer), llmod_id)
-                }
-            }
-        }).collect::<Vec<_>>();
+        for module in in_memory {
+            let buffer = ModuleBuffer::new(module.module_llvm.llmod());
+            let llmod_id = CString::new(&module.name[..]).unwrap();
+            serialized_modules.push((SerializedModule::Local(buffer), llmod_id));
+        }
         // Sort the modules to ensure we produce deterministic results.
-        new_modules.sort_by(|module1, module2| module1.1.partial_cmp(&module2.1).unwrap());
-        serialized_modules.extend(new_modules);
-        serialized_modules.extend(cached_modules.into_iter().map(|(buffer, wp)| {
-            (buffer, CString::new(wp.cgu_name).unwrap())
-        }));
+        serialized_modules.sort_by(|module1, module2| module1.1.cmp(&module2.1));
 
         // For all serialized bitcode files we parse them and link them in as we did
         // above, this is all mostly handled in C++. Like above, though, we don't
@@ -850,7 +846,7 @@ fn module_name_to_str(c_str: &CStr) -> &str {
         bug!("Encountered non-utf8 LLVM module name `{}`: {}", c_str.to_string_lossy(), e))
 }
 
-fn parse_module<'a>(
+pub fn parse_module<'a>(
     cx: &'a llvm::Context,
     name: &CStr,
     data: &[u8],

--- a/src/librustc_codegen_llvm/lib.rs
+++ b/src/librustc_codegen_llvm/lib.rs
@@ -54,6 +54,7 @@ use syntax_pos::symbol::InternedString;
 pub use llvm_util::target_features;
 use std::any::Any;
 use std::sync::{mpsc, Arc};
+use std::ffi::CStr;
 
 use rustc::dep_graph::DepGraph;
 use rustc::middle::cstore::{EncodedMetadata, MetadataLoader};
@@ -386,13 +387,13 @@ impl ModuleLlvm {
 
     fn parse(
         cgcx: &CodegenContext<LlvmCodegenBackend>,
-        name: &str,
-        buffer: &back::lto::ModuleBuffer,
+        name: &CStr,
+        buffer: &[u8],
         handler: &Handler,
     ) -> Result<Self, FatalError> {
         unsafe {
             let llcx = llvm::LLVMRustContextCreate(cgcx.fewer_names);
-            let llmod_raw = buffer.parse(name, llcx, handler)?;
+            let llmod_raw = back::lto::parse_module(llcx, name, buffer, handler)?;
             let tm = match (cgcx.tm_factory.0)() {
                 Ok(m) => m,
                 Err(e) => {

--- a/src/test/run-make-fulldeps/lto-empty/Makefile
+++ b/src/test/run-make-fulldeps/lto-empty/Makefile
@@ -1,0 +1,12 @@
+-include ../tools.mk
+
+all: cdylib-fat cdylib-thin
+
+cdylib-fat:
+	$(RUSTC) lib.rs -C lto=fat -C opt-level=3 -C incremental=$(TMPDIR)/inc-fat
+	$(RUSTC) lib.rs -C lto=fat -C opt-level=3 -C incremental=$(TMPDIR)/inc-fat
+
+cdylib-thin:
+	$(RUSTC) lib.rs -C lto=thin -C opt-level=3 -C incremental=$(TMPDIR)/inc-thin
+	$(RUSTC) lib.rs -C lto=thin -C opt-level=3 -C incremental=$(TMPDIR)/inc-thin
+

--- a/src/test/run-make-fulldeps/lto-empty/lib.rs
+++ b/src/test/run-make-fulldeps/lto-empty/lib.rs
@@ -1,0 +1,1 @@
+#![crate_type = "cdylib"]

--- a/src/test/ui/lto-duplicate-symbols.stderr
+++ b/src/test/ui/lto-duplicate-symbols.stderr
@@ -1,6 +1,6 @@
 warning: Linking globals named 'foo': symbol multiply defined!
 
-error: failed to load bc of "lto_duplicate_symbols1.3a1fbbbh-cgu.0": 
+error: failed to load bc of "lto_duplicate_symbols2.3a1fbbbh-cgu.0": 
 
 error: aborting due to previous error
 

